src/rez/package_order.py

```diff
@@ -40,11 +94,71 @@ def reorder(self, iterable, key=None):
         Returns:
             list: Reordered ``iterable``
         """
+        key = key or (lambda x: x)
+        package_name = self._get_package_name_from_iterable(iterable, key=key)
+        return sorted(iterable,
+                      key=lambda x: self.sort_key(package_name, key(x).version),
+                      reverse=True)
+
+    @staticmethod
+    def _get_package_name_from_iterable(iterable, key=None):
+        """Utility method for getting a package from an iterable"""
+        try:
+            item = next(iter(iterable))
+        except (TypeError, StopIteration):
+            return None
+
+        key = key or (lambda x: x)
+        return key(item).name
+
+    def sort_key(self, package_name, version_like):
+        """Returns a sort key usable for sorting packages within the same family
+
+        Args:
+            package_name: (str) The family name of the package we are sorting
+            version_like: (Version|_LowerBound|_UpperBound|_Bound|VersionRange)
+                the version-like object you wish to generate a key for
+
+        Returns:
+            Comparable object
```

Comment:
From what I see, the return value is a tuple in most cases, except when `version_like` is of type `Version`? If so, it's a little weird that it sometimes return a tuple and sometimes a Comparable instance.

Reply:
Sorry, that docstring should probably read "Returns: Sortable object" as in some object that implements `__lt__`.  I would add it as a `Protocol`, but those were added into `typing` in python 3.8 and I'm sure we don't want to add a dependency on `typing_extensions` for just this and just in python 3.7.

The class would look something like this:
``` python
from typing import Protocol, TypeVar, runtime_checkable

@runtime_checkable
class SortableProtocol(Protocol):
    """A protocol to which all objects that can be sorted must adhere.
    `sorted()` uses the < operator to determine sort order, so anything
    that can be sorted must implement __lt__"""

    def __lt__(self, other: Any) -> bool:
        pass
```

Then we'd add a `SortableT = TypeVar("SortableT", bound=SortableProtocol)`

And then this class would return a `SortableT`

Reply:
I've added some clarifying language